Source/Plugins/Core/com.equella.core/src/com/tle/core/viewcount/service/ViewCountService.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to The Apereo Foundation under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * The Apereo Foundation licenses this file to you under the Apache License,
+ * Version 2.0, (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.tle.core.viewcount.service;
+
+import com.tle.beans.Institution;
+import com.tle.beans.entity.itemdef.ItemDefinition;
+import com.tle.beans.item.ItemKey;
+import com.tle.beans.viewcount.ViewcountAttachment;
+import com.tle.beans.viewcount.ViewcountItem;
+import java.util.List;
+
+public interface ViewCountService {
+  /**
+   * Get the view count of an Item.
+   *
+   * @param itemKey The Item's unique key.
+   */
+  int getItemViewCount(ItemKey itemKey);
+  /**
+   * Get the view count of an Attachment.
+   *
+   * @param itemKey The Item's unique key.
+   * @param attachmentUuid UUID of the Attachment.
+   */
+  int getAttachmentViewCount(ItemKey itemKey, String attachmentUuid);
+  /**
+   * Get all the entity {@link ViewcountItem} of an Institution.
+   *
+   * @param institution The Institution from where to get a list of `ViewcountItem`.
+   */
+  List<ViewcountItem> getItemViewCountList(Institution institution);
+  /**
+   * Get all the entity `{@link ViewcountAttachment} of an Institution.
+   *
+   * @param institution The Institution from where to get a list of `ViewcountAttachment`.
+   */
+  List<ViewcountAttachment> getAttachmentViewCountList(Institution institution, ItemKey itemKey);
+  /**
+   * The total Item view count of a Collection.
+   *
+   * @param col The collection for which to get the total Item view count.
+   */
+  int getItemViewCountForCollection(ItemDefinition col);
+  /**
+   * The total Attachment view count of a Collection.
+   *
+   * @param col The collection for which to get the total Attachment view count.
+   */
+  int getAttachmentViewCountForCollection(ItemDefinition col);
+  /**
+   * Add or update an instance of {@link ViewcountItem} for an Item.
+   *
+   * @param viewcountItem The instance of ViewcountItem to be updated.
+   */
+  void setItemViewCount(ViewcountItem viewcountItem);
```

Comment:
Being that this is a service layer, should it be more along the lines of:

```suggestion
  void setItemViewCount(ItemKey item, int viewCount);
```

Reply:
I have updated this.

Reply:
Ah I can see the issue I used to have.  With the signature of `(ItemKey item, int viewCount)`, we need to create a new instance of `ViewcountItem`, and then calling `dao.save` will trigger this exception.
`DuplicateKeyException: A different object with the same identifier value was already associated with the session`

Reply:
The reason is `setAttachmentViewCount` is called within `incrementAttachmentViewCount` which already fined the object by ID.

Reply:
Let's have some discussions for this.